File: Add Code Here/JAVA/0-1Knapsack.java
import java.util.*;
public class Solution
{
    public static int maxsteal(int index,int bagweight,int[] weight, int[] value,int[][]dp)
    {
        if(index==0)
        {
            if(weight[0]<=bagweight)
            {
                return value[0];
            }
            else
            {
                return 0;
            }
        }
        if(dp[index][bagweight]!=-1)
        {
            return dp[index][bagweight];
        }
        int take=Integer.MIN_VALUE;
        int nottake=0+maxsteal(index-1,bagweight,weight,value,dp);
        
        //Now check conditions for take
        if(weight[index]<=bagweight)
        {
            take=value[index]+maxsteal(index-1,bagweight-weight[index],weight,value,dp);
        }
        
        return dp[index][bagweight]=Math.max(take,nottake);
    }
    static int knapsack(int[] weight, int[] value, int n, int maxWeight) 
    {
        //Pick and Dont pick Concept
        //DP Approach
        //Since we have two parameters changing
        int[][]dp=new int[n][maxWeight+1];
        for (int i = 0; i < n; i++) 
        {
            Arrays.fill(dp[i], -1);
        }
        return maxsteal(n-1,maxWeight,weight,value,dp);

    }
}

File: Hacktoberfest2023-main/Add Code Here/C++/Dynamic_programming/Medium/01 Knapsack.cpp
#include <bits/stdc++.h>

using namespace std;

int knapSackRec(int W, int wt[], int val[], int index, int** dp)
{
	if (index < 0)
		return 0;
	if (dp[index][W] != -1)
		return dp[index][W];

	if (wt[index] > W) {
		dp[index][W] = knapSackRec(W, wt, val, index - 1, dp);
		return dp[index][W];
	}
	else {
		dp[index][W] = max(val[index]
						+ knapSackRec(W - wt[index], wt, val,
										index - 1, dp),
					knapSackRec(W, wt, val, index - 1, dp));

		return dp[index][W];
	}
}

int knapSack(int W, int wt[], int val[], int n)
{
	int** dp;
	dp = new int*[n];
	for (int i = 0; i < n; i++)
		dp[i] = new int[W + 1];
	for (int i = 0; i < n; i++)
		for (int j = 0; j < W + 1; j++)
			dp[i][j] = -1;
	return knapSackRec(W, wt, val, n - 1, dp);
}

int main()
{
	int profit[] = { 60, 100, 120 };
	int weight[] = { 10, 20, 30 };
	int W = 50;
	int n = sizeof(profit) / sizeof(profit[0]);
	cout << knapSack(W, weight, profit, n);
	return 0;
}
